File: go/ch02/linear_search.go
package ch02

// LinearSearch searches the needle in the haystack and returns the index, or
// -1, if the needle wasn't found.
func LinearSearch[T comparable](haystack []T, needle T) int {
	for i, e := range haystack {
		if e == needle {
			return i
		}
	}
	return -1
}

File: erlang/ch02/linear_search.erl
-module(linear_search).
-export([search/2]).

search([], _) -> -1;
search(Haystack, Needle) -> search(Haystack, Needle, 0).

search([], _, _) -> -1;
search([H|_], H, Acc) -> Acc;
search([_|T], H, Acc) -> search(T, H, Acc + 1).
